x-pack/plugins/security/server/session_management/session.ts

```diff
@@ -0,0 +1,405 @@
+        await this.options.sessionCookie.clear(request);
+        return null;
+      }
+
+      sessionValue.metadata.index = sessionIndexValue;
+    }
+
+    await this.options.sessionCookie.set(request, {
+      ...sessionCookieValue,
+      ...sessionExpirationInfo,
+    });
+
+    sessionLogger.debug('Successfully extended existing session.');
+
+    return { ...sessionValue, ...sessionExpirationInfo } as Readonly<SessionValue>;
+  }
+
+  /**
+   * Clears session value for the specified request.
+   * @param request Request instance to clear session value for.
+   */
+  async clear(request: KibanaRequest) {
+    const sessionCookieValue = await this.options.sessionCookie.get(request);
+    if (!sessionCookieValue) {
+      return;
+    }
+
+    const sessionLogger = this.getLoggerForSID(sessionCookieValue.sid);
+    sessionLogger.debug('Invalidating session.');
+
+    await Promise.all([
+      this.options.sessionCookie.clear(request),
+      this.options.sessionIndex.clear(sessionCookieValue.sid),
+    ]);
+
+    sessionLogger.debug('Successfully invalidated session.');
+  }
+
+  private calculateExpiry(
+    currentLifespanExpiration?: number | null
+  ): { idleTimeoutExpiration: number | null; lifespanExpiration: number | null } {
+    const now = Date.now();
+    // if we are renewing an existing session, use its `lifespanExpiration` -- otherwise, set this value
+    // based on the configured server `lifespan`.
+    // note, if the server had a `lifespan` set and then removes it, remove `lifespanExpiration` on renewed sessions
+    // also, if the server did not have a `lifespan` set and then adds it, add `lifespanExpiration` on renewed sessions
+    const lifespanExpiration =
+      currentLifespanExpiration && this.lifespan
+        ? currentLifespanExpiration
+        : this.lifespan && now + this.lifespan.asMilliseconds();
+    const idleTimeoutExpiration = this.idleTimeout && now + this.idleTimeout.asMilliseconds();
+
+    return { idleTimeoutExpiration, lifespanExpiration };
+  }
+
+  /**
+   * Converts value retrieved from the index to the value returned to the API consumers.
+   * @param sessionIndexValue The value returned from the index.
+   * @param decryptedContent Decrypted session value content.
+   */
+  private static sessionIndexValueToSessionValue(
+    sessionIndexValue: Readonly<SessionIndexValue>,
+    { username, state }: SessionValueContentToEncrypt
+  ): Readonly<SessionValue> {
+    // Extract values that are specific to session index value.
+    const { usernameHash, content, ...publicSessionValue } = sessionIndexValue;
+    return { ...publicSessionValue, username, state, metadata: { index: sessionIndexValue } };
+  }
+
+  /**
+   * Creates logger scoped to a specified session ID.
+   * @param sid Session ID to create logger for.
+   */
+  private getLoggerForSID(sid: string) {
+    return this.options.logger.get(sid?.slice(-10));
```

Comment:
__note:__ having `?` here isn't technically necessary, but it'll help us in case we have bugs and `sid` happens to be not-string.